Tutorial pt 2 - CLI in depth (#713)
canonical/ubuntu-server-documentation: docs/tutorial/cli-in-depth.md

The `-l` flag tells `wc` to count lines only. When we use `<`, the file name does not appear in the output -- `wc` sees the file's contents as if they had been typed at the keyboard.

### Here-documents

There is a related form called a **here-document**, written as `<<`. It lets us type multi-line input directly in the terminal, ending when we type a delimiter we choose ourselves (often `EOF` is used as this marker -- short for "end of file"). Let's test this by sending a greeting into a new file called `greeting.txt`:

```{terminal}
:copy:
:user: ubuntu
:host: tutorial
:dir: ~
cat > greeting.txt << EOF
Hello from the terminal!
This text is going into a file.
EOF
```

The shell reads everything we type until it sees `EOF` on a line by itself, then passes it all as input to the command on the left. Let's confirm the file was created correctly:

```{terminal}
:copy:
:user: ubuntu
:host: tutorial
:dir: ~
cat greeting.txt

Hello from the terminal!
This text is going into a file.
```

## Pipes

Pipes are one of the most powerful ideas in the Unix philosophy: the output of one command becomes the input of the next. We connect them with the `|` operator (the "pipe" character).

> For example, `ls -la /usr/bin` produces many lines of output -- more than can fit on screen, and you can only see the end of the output (you can't scroll back to the top). To make it more convenient, we can pipe the output to `less` so we can page through it with the {kbd}`space` key:

Reviewer: ```suggestion
For example, `ls -la /usr/bin` produces many lines of output -- more than can fit on screen, and you can only see the end of the output (scrolling back up is very inconvenient, if you can even do it at all (some terminals can't!)). To make it more convenient, we can pipe the output to `less` so we can page through it with the {kbd}`space` key:
```

nit: i think most modern terminal emulators actually can scroll back up, so I reworded this

Author: good catch! adopted

```{terminal}
:copy:
:user: ubuntu
:host: tutorial
:dir: ~
ls -la /usr/bin | less
```

Press {kbd}`Q` to quit `less` when you are done.

We can also pipe to `grep` to filter the output, keeping only lines that match a pattern. Let's find all currently set environment variables whose names contain the word `LANG`:

```{terminal}
:copy:
:user: ubuntu
:host: tutorial
:dir: ~
env | grep LANG

LANG=C.UTF-8
```

Pipes can be chained together to build a **pipeline** -- each command passes its output along to the next. For example, to see the 3 most recently modified files in `/usr/bin`, we can sort by modification time and then cut the output down with `head`:

```{terminal}
:copy:
:user: ubuntu
:host: tutorial
:dir: ~
ls -lt /usr/bin | head -n 3

total 121540
-rwxr-xr-x 1 root root      174328 Apr 30 13:32 kmod
lrwxrwxrwx 1 root root           4 Apr 30 13:32 lsmod -> kmod
```

Each command in the pipeline only ever sees the output of the command immediately before it -- the shell connects them all behind the scenes.

## Chaining commands

> Pipes pass *data* from one command to another. Command chaining is different: it controls the *order* in which commands run, and whether the next command runs at all.

Reviewer: ```suggestion
Pipes pass *data* from one command to another and run all of their commands in parallel. This is to allow commands begin working on long streams of data while their source is still producing more. While often useful, you might not want that, in which case you would use **command chaining**.  Chaining is different: it controls the *order* in which commands run, and whether the next command runs at all.
```

Since we're talking about command order here, I think we should explicitly mention that pipe limbs run in parallel. 

I could be convinced that this warrants an even greater mention than this (since this has definitely caused me headache at least once when writing shell scripts), but for now I think mentioning it here is good.

Author: I actually didn't know they ran in parallel, so once again TIL :D 

I've adopted your wording, with some tweaks. I probably won't expand on it in this tutorial - if we ever write one about scripting, that would be the best place to expand on it imo. Maybe I'll do one for the new wiki?

The simplest way to chain commands is to separate them with a semicolon (`;`). The shell runs each command in turn, regardless of whether the previous one succeeded, with the output of each command shown on separate lines:

```{terminal}
:copy:
:user: ubuntu
:host: tutorial
:dir: ~
ls; pwd; whoami

errors.txt  file-list.txt  greeting.txt  output.txt
/home/ubuntu
ubuntu
```

> Often we want the second command to run *only* if the first one succeeded. We use `&&` for this:

Reviewer: ```suggestion
On Linux, the way processes in general declare whether they have "succeeded" or not is with an **exit code.** An exit code (also interchangeably called "exit status") is a number from zero to 255 that a process gives to its parent upon completion. "Success" is defined by exiting with code `0`, with *all* other values indicating (different types of) failure. These might be specific to the program, but in general your generic failure status will be exit code `1`. You can use `exit n` to immediately exit with code `n`.

Often we want the second command to run *only* if the first one succeeded. We use `&&` for this:
```

Since we're talking about "success" now, we should clarify what that means. There's more info later, but I think a few sentences here is fine. 

(tangent: git is awfully ambitious with where it's placing the diff on this particular comment :) )

Author: Didn't know this either - it seems like an interesting aside so I've called it out in an admonition block rather than just body text.

```{terminal}
:copy:
:user: ubuntu
:host: tutorial
:dir: ~
cd /tmp && ls

snap-private-tmp  systemd-private-[...]
```

If `cd /tmp` had failed (which can happen in a script when the directory does not exist), `ls` would not run at all.

The opposite of `&&` is `||`, which runs the second command only if the first one *failed*:

```{terminal}
:copy:
:user: ubuntu
:host: tutorial
:dir: ~/tmp
cd /nonexistent || echo "Could not change directory"

-bash: cd: /nonexistent: No such file or directory
Could not change directory
```

We can combine `&&` and `||` to produce a simple success-or-failure message:

```{terminal}
:copy:
:user: ubuntu
:host: tutorial
:dir: ~/tmp
ping -c1 localhost && echo "Host is up" || echo "Host is unreachable"

PING localhost (::1) 56 data bytes
[...]
Host is up
```

A few other operators are useful to know. Try each of these out, both with and without the operators, to see how they change the behaviour for yourself:

Background execution; `&`
: Runs a command in the background. The shell prompt returns immediately, and the command continues running, e.g.:
: ```text

>   nano file-list.txt &

Reviewer: `nano` is probably not the right choice of example for this, since a text editor is probably something you pretty much always want actually open, and backgrounding it doesn't really make sense. 

The example that immediately comes to mind is something like using `wget` to fetch a large file (say, an Ubuntu Server ISO 😉):
```
wget https://cdimage.ubuntu.com/ubuntu-server/resolute/daily-live/current/resolute-live-server-amd64.iso
```
but also I don't think having the example command fetch a >1G file is a good idea.  

`sleep` doesn't actually *do* anything, but it's probably still the most useful example. Maybe have them do `sleep 10`, observe that their shell prompt is gone for 10 seconds, then `sleep 10 &` and viola (🎻 )! shell remains accessible! 

In any case, it's probably worth mentioning `&` outputs the job number and PID of the backgrounded process:
```
grayson.wolf@canonical.com@penemue:~$ sleep 10 &
[1] 235741
```
if only so people know What The Weird Numbers Mean or if they need to `kill` it later.

Author: I use `gedit <file> &` all the time 😅 this is what happens when you get taught to Linux by *not* computer science people, you get bad habits

Sleep is a nice example though, thanks! Adding both that and the PID note.

  ```
  Would open the `file-list.txt` file in a text editor but still keep the terminal free to use.

Command grouping; `{}`
: Group scommands together so they share the same output redirection or run as a unit. E.g. to create a quick log file you could run:
: ```text
  { echo "System Report"; date; uptime; } > report.txt
  ```

Subshell execution; `()`
: Run commands in a subshell, where the commands are forgotten the moment they finish running. This is useful for isolating side effects like a `cd` -- the directory change does not affect the current shell.
: ```text
  (export NAME="Alice"; echo "Hello $NAME")
  echo "Name is: $NAME"

>   ```

Reviewer: I think this makes more sense to be moved into the section on env vars (maybe just calling it "The Environment" in general), because at least in my case the most common usage of this kind of thing is capturing output like `GITHUB_TOKEN="$(gh auth token)"`, and since the important part about `()` is that it doesn't change the environment (and since your example uses env vars!)

Author: Yeah this is a bit of a chicken-and-egg situation. I decided that since several operators are used in the env vars section (not just the subshell execution), they should be specifically introduced first as a logical/conceptual group so that a self-reinforcing loop is made when it's actually used in examples for useful things later, without necessarily coupling them together in the mind of the reader (implying more restricted usage than reality allows).

Line continuation; `\`
: Continue a long command on the next line, which helps a lot with command readability.

> The `()` subshell is particularly handy when we want to work in a different directory without losing our current location:

Reviewer: should this have a command with it? 

in any case it seems separated from the rest of the info by the line continuation bit, so this should move.

Author: Ah, no this is an artifact of my editing style. Anytime I'm not happy with the wording I redraft on multiple lines - sometimes I forget to delete the discarded ones lol. deleted it now.

:dir: ~
echo $?

2
```

This is especially useful in scripts, where we can check whether the previous command succeeded before deciding what to do next.

## Shell expansion

When we press {kbd}`Enter` on a command, the shell does not send our text directly to the program. First, it processes the command line and **expands** certain patterns into their final values. Understanding expansion explains a lot of seemingly "magic" shell behaviour.

### Tilde expansion

We have already used tilde expansion without thinking about it. When the shell sees `~` at the start of a word, it replaces it with the value of `$HOME`:

```{terminal}
:copy:
:user: ubuntu
:host: tutorial
:dir: ~
echo ~

/home/ubuntu
```

```{terminal}
:copy:
:user: ubuntu
:host: tutorial
:dir: ~
echo ~/my-file.txt

/home/ubuntu/my-file.txt
```

### Variable expansion

We have seen this already: `$VARIABLE_NAME` (or `${VARIABLE_NAME}`) is replaced with the variable's value before the command runs. The braces form is important when the variable name might be confused with the characters that follow it:

```{terminal}
:copy:
:user: ubuntu
:host: tutorial
:dir: ~
DAY="Monday"
echo "Today is ${DAY}!"

Today is Monday!
```

> Without the braces, the shell would try to expand `$DAY!` as a single variable name and find nothing.

Reviewer: ```suggestion
Without the braces, the shell would try to expand `$DAY!` as a single variable name and find nothing.

It's also important to note that this variable expansion applies *everywhere* in the line, even with the command name itself!:
\`\`\`{terminal}
:copy:
:user: ubuntu
:host: terminal
:dir: ~
CMD="echo"
$CMD "It's dynamic!"
It's dynamic!
\`\`\`
This can be useful in, say, a script that runs one of two very similar commands with the same argument, so you don't have to re-type everything twice in your conditional.
```

Added a brief caveat about using variables as command names (you might need to remove the anti-github-markdown backslashes).

Author: Added!
